File: packages/crypto/src/__tests__/group-ratchet.test.ts
import { describe, it, expect } from 'vitest';
import {
  createOutboundGroupSession,
  createInboundGroupSession,
  encryptGroupMessage,
  decryptGroupMessage,
  MAX_SKIPPED_GROUP_MESSAGE_KEYS,
} from '../group/ratchet';

const te = new TextEncoder();
const td = new TextDecoder();
const ad = te.encode('group-associated-data');

function joinAtCurrent(outbound: ReturnType<typeof createOutboundGroupSession>) {
  return createInboundGroupSession(outbound.sessionId, outbound.chainKey, outbound.counter);
}

describe('Group ratchet (Megolm-style, one-way per-sender chain)', () => {
  it('a member sends, another decrypts — first message of a brand-new session', () => {
    const outbound = createOutboundGroupSession();
    const inbound = joinAtCurrent(outbound);

    const { header, ciphertext } = encryptGroupMessage(outbound, te.encode('hello group'), ad);
    const plaintext = decryptGroupMessage(inbound, header, ciphertext, ad);

    expect(td.decode(plaintext)).toBe('hello group');
  });

  it('a run of messages from one sender stays in sync for a receiver who joined at counter 0', () => {
    const outbound = createOutboundGroupSession();
    const inbound = joinAtCurrent(outbound);
    const messages = ['one', 'two', 'three', 'four', 'five'];

    for (const text of messages) {
      const { header, ciphertext } = encryptGroupMessage(outbound, te.encode(text), ad);
      const plaintext = decryptGroupMessage(inbound, header, ciphertext, ad);
      expect(td.decode(plaintext)).toBe(text);
    }
  });

  it('handles out-of-order delivery (message 2 arrives before message 1)', () => {
    const outbound = createOutboundGroupSession();
    const inbound = joinAtCurrent(outbound);

    const msg1 = encryptGroupMessage(outbound, te.encode('first'), ad);
    const msg2 = encryptGroupMessage(outbound, te.encode('second'), ad);

    const plaintext2 = decryptGroupMessage(inbound, msg2.header, msg2.ciphertext, ad);
    expect(td.decode(plaintext2)).toBe('second');

    const plaintext1 = decryptGroupMessage(inbound, msg1.header, msg1.ciphertext, ad);
    expect(td.decode(plaintext1)).toBe('first');
  });

  it('a member who joins mid-conversation cannot decrypt history from before they joined (no retroactive access)', () => {
    const outbound = createOutboundGroupSession();

    // Two messages sent before the new member has any session for this sender.
    const before1 = encryptGroupMessage(outbound, te.encode('sent before joining #1'), ad);
    const before2 = encryptGroupMessage(outbound, te.encode('sent before joining #2'), ad);

    // The new member's inbound session starts at the sender's CURRENT counter —
    // exactly what a real group.key-share delivers (docs/13-roadmap.md).
    const lateJoiner = joinAtCurrent(outbound);

    expect(() => decryptGroupMessage(lateJoiner, before1.header, before1.ciphertext, ad)).toThrow();
    expect(() => decryptGroupMessage(lateJoiner, before2.header, before2.ciphertext, ad)).toThrow();

    // But a message sent after they joined decrypts fine.
    const after = encryptGroupMessage(outbound, te.encode('sent after joining'), ad);
    const plaintext = decryptGroupMessage(lateJoiner, after.header, after.ciphertext, ad);
    expect(td.decode(plaintext)).toBe('sent after joining');
  });

  it('epoch rotation: a removed member holding the old session cannot decrypt messages under the new one', () => {
    const oldOutbound = createOutboundGroupSession();
    const removedMember = joinAtCurrent(oldOutbound); // holds the pre-rotation session

    // Remaining members rotate to a brand-new outbound session (docs/05-crypto-architecture.md's
    // "on member removal" design) — the removed member never receives this one.
    const newOutbound = createOutboundGroupSession();
    const remainingMember = joinAtCurrent(newOutbound);

    const postRotation = encryptGroupMessage(newOutbound, te.encode('after removal'), ad);

    // The remaining member (who has the new session) decrypts fine...
    const plaintext = decryptGroupMessage(remainingMember, postRotation.header, postRotation.ciphertext, ad);
    expect(td.decode(plaintext)).toBe('after removal');

    // ...but the removed member's stale session is for a DIFFERENT sessionId entirely
    // and throws rather than producing garbage.
    expect(() => decryptGroupMessage(removedMember, postRotation.header, postRotation.ciphertext, ad)).toThrow(/different group session/);
  });

  it('forward secrecy: advancing the outbound chain means an earlier chain key cannot be recovered', () => {
    const outbound = createOutboundGroupSession();
    const staleChainKey = outbound.chainKey;

    encryptGroupMessage(outbound, te.encode('first'), ad);
    encryptGroupMessage(outbound, te.encode('second'), ad);

    expect(outbound.chainKey).not.toEqual(staleChainKey);
  });

  it('rejects a tampered ciphertext (bit flip) rather than returning corrupted plaintext', () => {
    const outbound = createOutboundGroupSession();
    const inbound = joinAtCurrent(outbound);
    const { header, ciphertext } = encryptGroupMessage(outbound, te.encode('sensitive'), ad);

    const tampered = new Uint8Array(ciphertext);
    tampered[0] = tampered[0]! ^ 0xff;

    expect(() => decryptGroupMessage(inbound, header, tampered, ad)).toThrow();
  });

  it('rejects a message re-encrypted under the wrong associated data (cross-group replay)', () => {
    const outbound = createOutboundGroupSession();
    const inbound = joinAtCurrent(outbound);
    const { header, ciphertext } = encryptGroupMessage(outbound, te.encode('for this group only'), ad);

    const wrongAd = te.encode('a different group entirely');
    expect(() => decryptGroupMessage(inbound, header, ciphertext, wrongAd)).toThrow();
  });

  it('rejects re-decrypting an already-consumed counter with no skip cache entry (replay/duplicate)', () => {
    const outbound = createOutboundGroupSession();
    const inbound = joinAtCurrent(outbound);
    const { header, ciphertext } = encryptGroupMessage(outbound, te.encode('once only'), ad);

    decryptGroupMessage(inbound, header, ciphertext, ad); // consumes counter 0
    expect(() => decryptGroupMessage(inbound, header, ciphertext, ad)).toThrow(/no longer available/);
  });

  it('bounds skipped-key derivation — refuses an absurd counter jump rather than allocating unboundedly (DoS guard)', () => {
    const outbound = createOutboundGroupSession();
    const inbound = joinAtCurrent(outbound);

    // Fabricate a header claiming a counter far beyond the bound, as if a malicious
    // sender (or corrupted delivery) tried to force a huge skip-key derivation.
    const { ciphertext } = encryptGroupMessage(outbound, te.encode('whatever'), ad);
    const maliciousHeader = { sessionId: outbound.sessionId, counter: MAX_SKIPPED_GROUP_MESSAGE_KEYS + 10 };

    expect(() => decryptGroupMessage(inbound, maliciousHeader, ciphertext, ad)).toThrow(/DoS guard/);
  });
});

File: apps/mobile/test/group/group_ratchet_test.dart
/// Self-consistency test suite mirroring
/// `packages/crypto/src/__tests__/group-ratchet.test.ts` — the group ratchet is a
/// zero-DH, one-way symmetric chain (see lib/crypto/group/ratchet.dart's docstring),
/// so unlike the 1:1 Double Ratchet its encrypt/decrypt functions take no internal
/// randomness at all once a session exists — every scenario below is a real,
/// deterministic exercise of the ported algorithm (round trip, out-of-order,
/// wrong-session rejection, forward secrecy, tamper detection).
library;

import 'dart:convert';
import 'dart:typed_data';
import 'package:flutter_test/flutter_test.dart';
import 'package:comm_mobile/crypto/group/ratchet.dart';

void main() {
  group('Group ratchet (Megolm-style)', () {
    test('round trip: outbound session encrypts, matching inbound session decrypts', () async {
      final outbound = createOutboundGroupSession();
      final ad = utf8.encode('group:conversation-id');

      // A real inbound session is created from the chain key value AS SHARED (before
      // any message is sent) — mirrored here by snapshotting it before encrypting,
      // since encryptGroupMessage mutates outbound.chainKey in place.
      final inbound = createInboundGroupSession(outbound.sessionId, outbound.chainKey, 0);

      final result = await encryptGroupMessage(outbound, utf8.encode('hello group'), ad);
      final plaintext = await decryptGroupMessage(inbound, result.header, result.ciphertext, ad);
      expect(utf8.decode(plaintext), 'hello group');
    });

    test('a full sequence of messages from one sender stays in sync', () async {
      final outbound = createOutboundGroupSession();
      final ad = utf8.encode('group:conversation-id');
      final inbound = createInboundGroupSession(outbound.sessionId, outbound.chainKey, 0);

      for (final text in ['one', 'two', 'three', 'four']) {
        final result = await encryptGroupMessage(outbound, utf8.encode(text), ad);
        final plaintext = await decryptGroupMessage(inbound, result.header, result.ciphertext, ad);
        expect(utf8.decode(plaintext), text);
      }
    });

    test('handles out-of-order delivery (message 2 arrives before message 1)', () async {
      final outbound = createOutboundGroupSession();
      final ad = utf8.encode('group:conversation-id');
      final inbound = createInboundGroupSession(outbound.sessionId, outbound.chainKey, 0);

      final msg1 = await encryptGroupMessage(outbound, utf8.encode('first'), ad);
      final msg2 = await encryptGroupMessage(outbound, utf8.encode('second'), ad);

      final plaintext2 = await decryptGroupMessage(inbound, msg2.header, msg2.ciphertext, ad);
      expect(utf8.decode(plaintext2), 'second');

      final plaintext1 = await decryptGroupMessage(inbound, msg1.header, msg1.ciphertext, ad);
      expect(utf8.decode(plaintext1), 'first');
    });

    test('rejects a header belonging to a different group session', () async {
      final outboundA = createOutboundGroupSession();
      final outboundB = createOutboundGroupSession();
      final ad = utf8.encode('group:conversation-id');
      final inboundForA = createInboundGroupSession(outboundA.sessionId, outboundA.chainKey, 0);

      final msgFromB = await encryptGroupMessage(outboundB, utf8.encode('wrong session'), ad);

      expect(
        () => decryptGroupMessage(inboundForA, msgFromB.header, msgFromB.ciphertext, ad),
        throwsA(isA<StateError>()),
      );
    });

    test('rejects a tampered ciphertext rather than returning corrupted plaintext', () async {
      final outbound = createOutboundGroupSession();
      final ad = utf8.encode('group:conversation-id');
      final inbound = createInboundGroupSession(outbound.sessionId, outbound.chainKey, 0);

      final result = await encryptGroupMessage(outbound, utf8.encode('sensitive'), ad);
      final tampered = Uint8List.fromList(result.ciphertext);
      tampered[0] = tampered[0] ^ 0xff;

      expect(() => decryptGroupMessage(inbound, result.header, tampered, ad), throwsA(anything));
    });

    test('forward secrecy: the chain key moves forward and cannot decrypt with a stale copy', () async {
      final outbound = createOutboundGroupSession();
      final ad = utf8.encode('group:conversation-id');
      final inbound = createInboundGroupSession(outbound.sessionId, outbound.chainKey, 0);

      final staleChainKey = Uint8List.fromList(inbound.chainKey);
      final msg = await encryptGroupMessage(outbound, utf8.encode('advance'), ad);
      await decryptGroupMessage(inbound, msg.header, msg.ciphertext, ad);

      expect(inbound.chainKey, isNot(equals(staleChainKey)));
    });

    test('a newly-added member (inbound session starting at a later counter) cannot decrypt earlier history', () async {
      final outbound = createOutboundGroupSession();
      final ad = utf8.encode('group:conversation-id');

      final msg1 = await encryptGroupMessage(outbound, utf8.encode('before the new member joined'), ad);
      // The new member's inbound session is created from the chain key AT THE
      // CURRENT counter (2, since one message already advanced it) — never from
      // counter 0, matching createInboundGroupSession's documented "no retroactive
      // history access" contract.
      final lateJoinerInbound = createInboundGroupSession(outbound.sessionId, outbound.chainKey, outbound.counter);

      expect(
        () => decryptGroupMessage(lateJoinerInbound, msg1.header, msg1.ciphertext, ad),
        throwsA(isA<StateError>()),
      );

      final msg2 = await encryptGroupMessage(outbound, utf8.encode('after the new member joined'), ad);
      final plaintext = await decryptGroupMessage(lateJoinerInbound, msg2.header, msg2.ciphertext, ad);
      expect(utf8.decode(plaintext), 'after the new member joined');
    });
  });
}
